### PR TITLE
Add basic auth

### PR DESCRIPTION
**Pull request by wltsmrz**

### CODE REVIEW

**Comment by wltsmrz** [comment]
Halt

File: src/core/server.js [hunk 1]
```diff
@@ -505,11 +513,13 @@ Server.prototype.connect = function() {
...
-        log.info(self.getServerID(), 'onerror:', e.data || e);
+        log.info(`${self.getServerID()} onerror: ${e.toString()}`);
       }
 
-      if (Server.TLS_ERRORS.indexOf(e.message) !== -1) {
-        // Unrecoverable
+      if (Server.TLS_ERRORS.includes(e.message)) {
+        throw e;
+      }
+      if (e.message.includes('unexpected')) {
```
Comment (wltsmrz) [inline src/core/server.js]: Not sure if this is going to be caught in the browser https://github.com/websockets/ws/blob/7debd827d39925650bba5a270811ed82616f0eb1/lib/WebSocket.js#L693

Comment (wltsmrz) [inline src/core/server.js]: Probably not

File: src/core/remote.js [hunk 2]
```diff
@@ -149,6 +149,9 @@ function Remote(options = {}) {
   if (!Array.isArray(this.servers)) {
     throw new TypeError('servers must be an array');
   }
+  if (!(_.isUndefined(this.basic_auth) || _.isString(this.basic_auth))) {
```
Comment (sublimator) [inline src/core/remote.js]: Why the check for undefined ?

File: src/core/server.js [hunk 3]
```diff
@@ -442,23 +442,24 @@ Server.prototype.connect = function() {
...
-    } catch (error) {
-      throw new Error('"proxy" option is not supported in the browser');
-    }
-    const agent = new HttpsProxyAgent(opts);
+  const wsOptions = {};
 
-    this._ws = new WebSocket(this._opts.url, {agent: agent});
-  } else {
-    this._ws = new WebSocket(this._opts.url);
+  if (!_.isUndefined(this._remote.proxy)) {
```
Comment (sublimator) [inline src/core/server.js]: Shame lodash doesn't have `isDefined `

File: src/core/remote.js (hunk 2 above)
Comment (wltsmrz) [inline src/core/remote.js]: basic_auth is optional, so it must either be undefined or a string

File: src/index.js [hunk 4]
```diff
@@ -1,4 +1,7 @@
 'use strict';
+
+require('babel-core/polyfill');
```
Comment (clark800) [inline src/index.js]: What's this for? Can we add a comment?

Comment (wltsmrz) [inline src/index.js]: Added comment

File: src/index.js [hunk 5]
```diff
@@ -1,4 +1,8 @@
 'use strict';
+
+// Enable core-js polyfills
```
Comment (clark800) [inline src/index.js]: I still don't know what this is for. Like what functionality are we using that makes us need this, what does it provide? Why was it working without this before?

Comment (wltsmrz) [inline src/index.js]: Updated comment

File: src/index.js [hunk 6]
```diff
@@ -1,4 +1,11 @@
 'use strict';
+
+/* eslint-disable max-len */
+// Enable core-js polyfills. This allows use of ES6/7 extensions listed here:
+// https://github.com/zloirock/core-js/blob/fb0890f32dabe8d4d88a4350d1b268446127132e/shim.js#L1-L103
```
Comment (clark800) [inline src/index.js]: Are you using one of these that we weren't using before?

Comment (clark800) [inline src/index.js]: I see: String.includes and Array.includes

**Comment by clark800** [comment]
LGTM

**Pull request merged**